src/telemetry/Counter.h

```diff
@@ -8,197 +8,163 @@
 
 #include "zeek/Span.h"
 #include "zeek/telemetry/MetricFamily.h"
+#include "zeek/telemetry/telemetry.bif.h"
 
-#include "broker/telemetry/fwd.hh"
+#include "opentelemetry/sdk/metrics/async_instruments.h"
+#include "opentelemetry/sdk/metrics/sync_instruments.h"
 
 namespace zeek::telemetry {
-
-class DblCounterFamily;
-class IntCounterFamily;
-class Manager;
-
-/**
- * A handle to a metric that represents an integer value that can only go up.
- */
-class IntCounter {
+template<typename BaseType>
+class BaseCounter {
 public:
-    friend class IntCounterFamily;
-
-    static inline const char* OpaqueName = "IntCounterMetricVal";
-
-    IntCounter() = delete;
-    IntCounter(const IntCounter&) noexcept = default;
-    IntCounter& operator=(const IntCounter&) noexcept = default;
+    using Handle = opentelemetry::nostd::shared_ptr<opentelemetry::metrics::Counter<BaseType>>;
 
     /**
      * Increments the value by 1.
      */
-    void Inc() noexcept { broker::telemetry::inc(hdl); }
+    void Inc() noexcept { Inc(1); }
 
     /**
      * Increments the value by @p amount.
      * @pre `amount >= 0`
      */
-    void Inc(int64_t amount) noexcept { broker::telemetry::inc(hdl, amount); }
+    void Inc(BaseType amount) noexcept {
+        handle->Add(amount, attributes);
```

Comment:
There is a lookup in the `GetOrAdd` methods that does a comparison across all of the existing counters for a matching set of labels. That's probably what it is, without having actually looked at the flamegraph. In some cases we do store the counters locally (`session::Manager`, for one) but we don't do that in every case.

Reply:
Ah, I took it as something within `handle->Add(amount, attributes)`. opentelemetry needs to internally find the right counter of the given labels/attributes.

Reply:
I'm looking at a flamegraph of it now. I might have to open an issue on their side, as this dives really quickly down into their code. `MetricAttributeIterable::ForEachKeyValue` is our code. I might be able to optimize the loop itself by storing the values locally as `opentelemetry::common::AttributeValue` objects instead of strings. That'd avoid a conversion every time we do a lookup, but `ForEachKeyValue` is dominated mostly by the call to the callback method passed as an argument. That callback function comes from OTEL itself so we don't have much control over its performance.

Reply:
https://github.com/zeek/zeek/pull/3448/commits/ec5b2abdfb6c98f5bd71e97db4dc167edba5524e cuts down on it a bit by storing a secondary map that are views into the existing string map, but in the otel types so it doesn't have to do conversions. There's still a performance hit from the function call though. I'll open a ticket on their side to see if they can do anything about it.

Reply:
I talked to one of the OTEL guys and shared a flamegraph with him. He agreed that it's a performance issue for it to be rehashing the map every time, even though the map hasn't changed at all since the last time. He said he'd look into it as part of some other performance issues he's investigating soon.